proposals/0000-hasfield-incoherence.rst

```diff
@@ -0,0 +1,279 @@
+as to when a user can declare custom ``HasField`` instances. In this proposal
+we relax these restrictions, allowing users to define ``HasField`` instances
+when they were previously unable to.
+
+
+Motivation
+----------
+
+The current conditions prohibit a user from defining ``HasField`` custom
+instances under a few conditions. This proposal argues that these constraints
+are excessively restrictive and that GHC would benefit from relaxing these
+constraints.
+
+Consider the following module::
+
+  module Reactive.Banana ( Behavior ) where
+  newtype Behavior a = Behavior { unB :: ... }
+
+This type represents a time-varying value, and occurs in the
+``reactive-banana`` project. A ``Behavior`` is a ``Functor``, and it would be
+nice if we could "lift" any fields in ``a`` to ``Behavior a``::
+
+  instance HasField x a b => HasField (x :: k) (Behavior a) (Behavior b) where
+    getField b = getField @x <$> b
+
+Unfortunately, because ``Behavior`` was defined as ``{ unB :: ... }``, GHC
+considers it to have fields, and we are unable to write the ``HasField``
+instance we want. Worse, this is an abstraction leak - users of
+``reactive-banana`` can only see ``data Behavior a`` - its constructor and
+fields are not exported. This leads to a confusing error message if a user
+tries to define the above custom (orphan) instance, as they are told the type
+has fields, though without reading the source they are otherwise unable to
+observe that.
+
+This was originally reported as `GHC issue #21324 <https://gitlab.haskell.org/ghc/ghc/-/issues/21324>`_.
+
+A similar request is at `issue #21369
+<https://gitlab.haskell.org/ghc/ghc/-/issues/21369>`_. In this issue,
+``@parsonsmatt`` would like to write::
+
+  instance
+      ( SymbolToField sym ent typ
+      , PersistEntity ent
+      )
+    =>
+      HasField sym (Entity ent) typ
+    where
+      getField ent =
+          view (persistFieldLens (symbolToField @sym @ent @typ)) ent
+
+However, they are unable to as ``Entity`` is defined to have fields. The
+workaround was to `redefine Entity without fields
+<https://github.com/yesodweb/persistent/pull/1381/files>`_, but this is a
+rather significant change, and if users wanted to pattern match on ``Entity``
+as they would before, they now need a ``MonadFail`` constraint (due to matching
+on a pattern synonym).
+
+If the constraints blocking this instance were dropped, Matt could simply have
+added the instance above.
+
+
+Proposed Change Specification
+-----------------------------
+
+In the `overloaded record fields proposal
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0023-overloaded-record-fields.rst>`_,
+there is a `set of limitations
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0023-overloaded-record-fields.rst#virtual-record-fields>`_
+as to when a user can declare custom ``HasField`` instances. These conditions
+are:
```

Comment:
For clarity (it was hard to figure out, at first, whether these where case where you were allowed or disallowed to write instances)
```suggestion
as to when a user can declare custom ``HasField`` instances. These conditions
prevent defining a ``HasField`` instance when:
```